File: src/NodeParser/UnknownTypeNodeParser.ts
import ts from "typescript";
import type { SubNodeParser } from "../SubNodeParser.js";
import type { BaseType } from "../Type/BaseType.js";
import { UnknownType } from "../Type/UnknownType.js";

export class UnknownTypeNodeParser implements SubNodeParser {
    public supportsNode(node: ts.KeywordTypeNode): boolean {
        return node.kind === ts.SyntaxKind.UnknownKeyword;
    }

    public createType(): BaseType {
        return new UnknownType(false);
    }
}

File: internal/parser/unknown_type_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// UnknownTypeNodeParser handles the `unknown` keyword
// (src/NodeParser/UnknownTypeNodeParser.ts).
type UnknownTypeNodeParser struct{}

func NewUnknownTypeNodeParser() *UnknownTypeNodeParser {
	return &UnknownTypeNodeParser{}
}

func (p *UnknownTypeNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindUnknownKeyword
}

func (p *UnknownTypeNodeParser) CreateType(_ *ast.Node, _ *Context, _ *types.ReferenceType) types.Type {
	return &types.UnknownType{ErroredSource: false}
}
